File: src/server/protocol/socketHandshake.ml
(**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

(* When a client connects over the socket, there is a handshake. The client writes a message
 * and the Flow server monitor responds *)

type build_id = string

let build_revision =
  match Build_id.build_revision with
  | "" -> Flow_version.version
  | x -> x

(*
 * Handshake
 * 1. client sends (string*string) where fst is a JSON string that can deserialized
 *    in all cases and includes client_build_id, and snd is a Marshal.to_string that will
 *    only be deserialized if the client and server have the same build_id
 * 2. server sends back (string*string option) where fst is a JSON string that can be
 *    deserialized in all cases and says the server_build_id plus the server
 *    intent, and snd is either Some (Marshal.to_string _) if the server is sure
 *    that the client will be able to deserialize it thanks to build ids matching,
 *    or is None.
 * 3. Server behavior is as signalled by 'server_intent' in step 2:
 *    - server might intend to exit now (e.g. upon stop request, or mismatch)
 *    - might persist but will hang up the connection now (e.g. mismatch, initializing)
 *    - might persist and leave the connection open (e.g. ok)
 *)

type client_handshake_wire = string * string

type server_handshake_wire = string * string option

(* What to do on a version mismatch *)
type version_mismatch_strategy =
  | Always_stop_server (* Always stop the server *)
  | Stop_server_if_older (* Stop the server if it is older than the client. Otherwise send an error *)
  | Error_client

(* Don't stop the server, just send an error to the client *)

type client_to_monitor_1 = {
  client_build_id: build_id;
  (* Build_id.build_revision for the client *)
  client_version: string;
  (* Flow_version.version for the client *)
  is_stop_request: bool;
  (* are we requesting the server to stop? *)
  server_should_hangup_if_still_initializing: bool;
  version_mismatch_strategy: version_mismatch_strategy; (* What to do on a version mismatch *)
}

type server_intent =
  | Server_will_exit (* after receiving a stop request or as a result of a version mismatch *)
  | Server_will_hangup (* version mismatch but neither client nor server should restart *)
  | Server_will_continue

(* upon success *)

type monitor_to_client_1 = {
  server_build_id: build_id;
  (* Build_id.build_revision for the server *)
  server_bin: string;
  (* filepath to the server binary *)
  server_intent: server_intent;
  (* The result of the handshake *)
  server_version: string; (* Flow_version.version for the server *)
}

type client_type =
  | Ephemeral (* a client that sends a request, gets a response, and disconnects *)
  | Persistent of {
      logging_context: FlowEventLogger.logging_context;
      lsp: Lsp.Initialize.params option;
    }

type client_to_monitor_2 = { client_type: client_type }

type monitor_to_client_2 =
  | Server_has_too_many_clients
  | Server_still_initializing of (ServerStatus.status * FileWatcherStatus.status)
  | Server_ready

type client_handshake = client_to_monitor_1 * client_to_monitor_2

type server_handshake = monitor_to_client_1 * monitor_to_client_2 option

let version_mismatch_strategy_to_string = function
  | Always_stop_server -> "Always_stop_server"
  | Stop_server_if_older -> "Stop_server_if_older"
  | Error_client -> "Error_client"

let string_to_version_mismatch_strategy = function
  | "Always_stop_server" -> Always_stop_server
  | "Stop_server_if_older" -> Stop_server_if_older
  | "Error_client" -> Error_client
  | _ ->
    (* If someone adds a new strategy in the future old servers shouldn't explode. *)
    Error_client

let client_to_monitor_1__to_json (c : client_to_monitor_1) : Hh_json.json =
  Hh_json.(
    JSON_Object
      [ ("client_build_id", JSON_String c.client_build_id);
        ("is_stop_request", JSON_Bool c.is_stop_request);
        ( "server_should_hangup_if_still_initializing",
          JSON_Bool c.server_should_hangup_if_still_initializing );
        ("client_version", JSON_String c.client_version);
        ( "version_mismatch_strategy",
          JSON_String (version_mismatch_strategy_to_string c.version_mismatch_strategy) );
        (* Deprecated - sent to talk to old servers *)
        ( "server_should_exit_if_version_mismatch",
          JSON_Bool
            (match c.version_mismatch_strategy with
            | Always_stop_server -> true
            | Stop_server_if_older -> true
            (* Any server reading this field is older than this client *)
            | Error_client -> false) ) ])

let default_client_to_monitor_1 =
  {
    client_build_id = "INCOMPATIBLE";
    is_stop_request = false;
    server_should_hangup_if_still_initializing = false;
    client_version = "0.0.0";
    version_mismatch_strategy = Error_client;
  }

let json_to__client_to_monitor_1 (json : Hh_json.json) : client_to_monitor_1 =
  Hh_json_helpers.(
    let json = Some json in
    let d = default_client_to_monitor_1 in
    let client_build_id = Jget.string_d json "client_build_id" ~default:d.client_build_id in
    let is_stop_request = Jget.bool_d json "is_stop_request" ~default:d.is_stop_request in
    let server_should_hangup_if_still_initializing =
      Jget.bool_d
        json
        "server_should_hangup_if_still_initializing"
        ~default:d.server_should_hangup_if_still_initializing
    in
    let client_version = Jget.string_d json "client_version" ~default:d.client_version in
    let version_mismatch_strategy =
      match Jget.string_opt json "version_mismatch_strategy" with
      | Some strategy -> string_to_version_mismatch_strategy strategy
      | None ->
        (match Jget.bool_opt json "server_should_exit_if_version_mismatch" with
        | Some true -> Always_stop_server
        | Some false
        | None ->
          Error_client)
    in
    {
      client_build_id;
      is_stop_request;
      server_should_hangup_if_still_initializing;
      client_version;
      version_mismatch_strategy;
    })

let monitor_to_client_1__to_json (m : monitor_to_client_1) : Hh_json.json =
  Hh_json.(
    let intent_to_string intent =
      match intent with
      | Server_will_exit -> "Server_will_exit"
      | Server_will_hangup -> "Server_will_hangup"
      | Server_will_continue -> "Server_will_continue"
    in
    JSON_Object
      [ ("server_build_id", JSON_String m.server_build_id);
        ("server_bin", JSON_String m.server_bin);
        ("server_intent", JSON_String (m.server_intent |> intent_to_string));
        ("server_version", JSON_String m.server_version) ])

let json_to__monitor_to_client_1 (json : Hh_json.json) : monitor_to_client_1 =
  Hh_json_helpers.(
    let json = Some json in
    let string_to_intent s =
      match s with
      | "Server_will_exit" -> Server_will_exit
      | "Server_will_hangup" -> Server_will_hangup
      | "Server_will_continue" -> Server_will_continue
      | _ -> raise (Jget.Parse ("unknown intent " ^ s))
    in
    let server_build_id = Jget.string_exn json "server_build_id" in
    let server_bin = Jget.string_exn json "server_bin" in
    let server_intent = Jget.string_exn json "server_intent" |> string_to_intent in
    let server_version = Jget.string_d json ~default:"0.0.0" "server_version" in
    { server_build_id; server_bin; server_intent; server_version })

File: rust_port/crates/flow_server_env/src/socket_handshake.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use lsp_types::InitializeParams;
use serde::Deserialize;
use serde::Serialize;

use crate::file_watcher_status;
use crate::server_status;

pub type BuildId = String;

pub fn build_revision() -> String {
    flow_common::flow_version::VERSION.to_string()
}

pub type ClientHandshakeWire = (String, Vec<u8>);

pub type ServerHandshakeWire = (String, Option<Vec<u8>>);

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum VersionMismatchStrategy {
    AlwaysStopServer,
    StopServerIfOlder,
    ErrorClient,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClientToMonitor1 {
    pub client_build_id: BuildId,
    pub client_version: String,
    pub is_stop_request: bool,
    pub server_should_hangup_if_still_initializing: bool,
    pub version_mismatch_strategy: VersionMismatchStrategy,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum ServerIntent {
    ServerWillExit,
    ServerWillHangup,
    ServerWillContinue,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MonitorToClient1 {
    pub server_build_id: BuildId,
    pub server_bin: String,
    pub server_intent: ServerIntent,
    pub server_version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ClientType {
    Ephemeral,
    Persistent { lsp_init_params: InitializeParams },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClientToMonitor2 {
    pub client_type: ClientType,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MonitorToClient2 {
    ServerHasTooManyClients,
    ServerStillInitializing(server_status::Status, file_watcher_status::Status),
    ServerReady,
}

pub type ClientHandshake = (ClientToMonitor1, ClientToMonitor2);

pub type ServerHandshake = (MonitorToClient1, Option<MonitorToClient2>);

pub fn version_mismatch_strategy_to_string(strategy: &VersionMismatchStrategy) -> &'static str {
    match strategy {
        VersionMismatchStrategy::AlwaysStopServer => "Always_stop_server",
        VersionMismatchStrategy::StopServerIfOlder => "Stop_server_if_older",
        VersionMismatchStrategy::ErrorClient => "Error_client",
    }
}

pub fn string_to_version_mismatch_strategy(s: &str) -> VersionMismatchStrategy {
    match s {
        "Always_stop_server" => VersionMismatchStrategy::AlwaysStopServer,
        "Stop_server_if_older" => VersionMismatchStrategy::StopServerIfOlder,
        "Error_client" => VersionMismatchStrategy::ErrorClient,
        _ => VersionMismatchStrategy::ErrorClient,
    }
}

pub fn client_to_monitor_1_to_json(c: &ClientToMonitor1) -> serde_json::Value {
    serde_json::json!({
        "client_build_id": c.client_build_id,
        "is_stop_request": c.is_stop_request,
        "server_should_hangup_if_still_initializing": c.server_should_hangup_if_still_initializing,
        "client_version": c.client_version,
        "version_mismatch_strategy": version_mismatch_strategy_to_string(&c.version_mismatch_strategy),
        "server_should_exit_if_version_mismatch": match c.version_mismatch_strategy {
            VersionMismatchStrategy::AlwaysStopServer => true,
            VersionMismatchStrategy::StopServerIfOlder => true,
            VersionMismatchStrategy::ErrorClient => false,
        },
    })
}

pub fn default_client_to_monitor_1() -> ClientToMonitor1 {
    ClientToMonitor1 {
        client_build_id: "INCOMPATIBLE".to_string(),
        is_stop_request: false,
        server_should_hangup_if_still_initializing: false,
        client_version: "0.0.0".to_string(),
        version_mismatch_strategy: VersionMismatchStrategy::ErrorClient,
    }
}

pub fn json_to_client_to_monitor_1(json: &serde_json::Value) -> ClientToMonitor1 {
    let d = default_client_to_monitor_1();
    let client_build_id = json
        .get("client_build_id")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or(d.client_build_id);
    let is_stop_request = json
        .get("is_stop_request")
        .and_then(|v| v.as_bool())
        .unwrap_or(d.is_stop_request);
    let server_should_hangup_if_still_initializing = json
        .get("server_should_hangup_if_still_initializing")
        .and_then(|v| v.as_bool())
        .unwrap_or(d.server_should_hangup_if_still_initializing);
    let client_version = json
        .get("client_version")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or(d.client_version);
    let version_mismatch_strategy = match json
        .get("version_mismatch_strategy")
        .and_then(|v| v.as_str())
    {
        Some(strategy) => string_to_version_mismatch_strategy(strategy),
        None => {
            match json
                .get("server_should_exit_if_version_mismatch")
                .and_then(|v| v.as_bool())
            {
                Some(true) => VersionMismatchStrategy::AlwaysStopServer,
                Some(false) | None => VersionMismatchStrategy::ErrorClient,
            }
        }
    };
    ClientToMonitor1 {
        client_build_id,
        is_stop_request,
        server_should_hangup_if_still_initializing,
        client_version,
        version_mismatch_strategy,
    }
}

pub fn monitor_to_client_1_to_json(m: &MonitorToClient1) -> serde_json::Value {
    let intent_to_string = |intent: &ServerIntent| -> &'static str {
        match intent {
            ServerIntent::ServerWillExit => "Server_will_exit",
            ServerIntent::ServerWillHangup => "Server_will_hangup",
            ServerIntent::ServerWillContinue => "Server_will_continue",
        }
    };
    serde_json::json!({
        "server_build_id": m.server_build_id,
        "server_bin": m.server_bin,
        "server_intent": intent_to_string(&m.server_intent),
        "server_version": m.server_version,
    })
}

pub fn json_to_monitor_to_client_1(json: &serde_json::Value) -> Result<MonitorToClient1, String> {
    let string_to_intent = |s: &str| -> Result<ServerIntent, String> {
        match s {
            "Server_will_exit" => Ok(ServerIntent::ServerWillExit),
            "Server_will_hangup" => Ok(ServerIntent::ServerWillHangup),
            "Server_will_continue" => Ok(ServerIntent::ServerWillContinue),
            _ => Err(format!("unknown intent {}", s)),
        }
    };
    let server_build_id = json
        .get("server_build_id")
        .and_then(|v| v.as_str())
        .ok_or_else(|| "missing server_build_id".to_string())?
        .to_string();
    let server_bin = json
        .get("server_bin")
        .and_then(|v| v.as_str())
        .ok_or_else(|| "missing server_bin".to_string())?
        .to_string();
    let server_intent = string_to_intent(
        json.get("server_intent")
            .and_then(|v| v.as_str())
            .ok_or_else(|| "missing server_intent".to_string())?,
    )?;
    let server_version = json
        .get("server_version")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or_else(|| "0.0.0".to_string());
    Ok(MonitorToClient1 {
        server_build_id,
        server_bin,
        server_intent,
        server_version,
    })
}
